#1084. GOcean Argument datatypes and PSyIR access expressions

sergisiso opened the pull request:
~~The Gocean example 8 PSy-layer can be fully generated using the PSyIR Fortran backend.~~ I spitted the original PR in 2, now it needs #1087 as well.

sergisiso commented:
@arporter or @rupertford This is ready for the first review.

arporter commented on src/psyclone/gocean1p0.py:
```diff
@@ -1947,12 +1986,32 @@ def __init__(self, arg):
+        # Get aggregate grid type accessors without the base name
+        access = self.dereference(base_field).split('%')[1:]
+
+        # Contstruct the PSyIR reference
```
"Construct"

sergisiso replied:
Done

arporter commented on src/psyclone/gocean1p0.py:
```diff
@@ -57,18 +57,19 @@
 from psyclone.psyir.symbols import SymbolTable, ScalarType, ArrayType, \
     INTEGER_TYPE, DataSymbol, ArgumentInterface, RoutineSymbol, Symbol, \
-    ContainerSymbol, DeferredType, TypeSymbol, GlobalInterface
+    ContainerSymbol, DeferredType, TypeSymbol, GlobalInterface, \
```
pylint complains that Symbol and GlobalInterface are not used.

sergisiso replied:
Deleted

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -49,7 +49,8 @@
+    Symbol, ContainerSymbol, GlobalInterface, INTEGER_TYPE, BOOLEAN_TYPE, \
+    ArgumentInterface, DeferredType
 from psyclone.psyir.symbols.datatypes import UnknownFortranType
 from psyclone.psyir.nodes import Node, Schedule, Loop, Statement, Container, \
```
pylint complains that Reference is not used.

sergisiso replied:
Deleted

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3708,9 +3704,59 @@ def __init__(self, call, arg_info, access):
+                    argument_access = ArgumentInterface.Access.READWRITE
+
+                # Find the tag or create a new symbol with expected attributes
+                new_argument = self._call.root.symbol_table.symbol_from_tag(
```
Presumably `root` gives the Schedule of the Invoke?

sergisiso replied:
yes, I made it more explicit now with `symtab = self._call.ancestor(InvokeSchedule).symbol_table` just in case we add a top module at some point.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3708,9 +3704,59 @@ def __init__(self, call, arg_info, access):
+        :raises NotImplementedError: this is an abstract method.
+
+        '''
+        raise NotImplementedError("Abstract method. Implement in the API.")
```
I think you can remove this line (and associated bit in the docstring) as you've marked it as an abstractmethod so it cannot be called.

sergisiso replied:
deleted

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3708,9 +3704,59 @@ def __init__(self, call, arg_info, access):
+        '''
+        raise NotImplementedError("Abstract method. Implement in the API.")
+
+    def infere_datatype(self):
```
"infer" and s/latter/later/ below.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/gocean1p0_test.py:
```diff
@@ -1722,6 +1723,117 @@ def test_gokernelarguments_append():
 
 
+def test_gokernelargument_infere_datatype():
+    ''' Check the GOcean specialisation of the infere_datatype work for each
```
"infer" and "works"

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -1056,6 +1064,20 @@ def test_invalid_reprod_pad_size(monkeypatch, dist_mem):
+    assert isinstance(arg.infere_datatype(), DeferredType)
+
+
+def test_argument_psyir_expression():
```
You can probably delete this test.

sergisiso replied:
deleted

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -58,10 +58,11 @@
     OMPParallelDirective, OMPDoDirective, OMPDirective, Directive, \
     ACCEnterDataDirective, ACCKernelsDirective, HaloExchange, Invoke, \
-    DataAccess, Kern, Arguments, CodedKern
+    DataAccess, Kern, Arguments, CodedKern, Argument
```
pylint complains that Reference is unused.

sergisiso replied:
Deleted

arporter reviewed (changes requested):
Looks good Sergi. I note that the Fortran generated for eg8 isn't *quite* complete - we don't declare `integer :: i`.
I think your choice of name `psyir_expression` is better than my `_grid_property_reference` although I realise they do slightly different things. Would it make sense for you to rename `_grid_property_reference` to `_grid_property_psyir_expression` (or similar) do you think?
Tests and examples are all OK with compilation.
I've checked for pylint and pycodestyle - see inline.
This is 'internal' development still so no need to update documentation.
Please could you add any touched classes to an `__all__` at the end of gocean1p0.

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1886,6 +1866,65 @@ def __init__(self, arg, arg_info, call):
+                            " 'field' or 'scalar', but found '{0}'."
+                            "".format(self.argument_type))
+
+    def infere_datatype(self):
```
I made this typo all over the place 🤦‍♂️ Fixed now.

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1886,6 +1866,65 @@ def __init__(self, arg, arg_info, call):
+
+        :raises InternalError: if this Argument type is not "field" or \
+                               "scalar".
+        :raises InternalError: if this argument is scalar but its space is \
```
I changed it to "space property" as it comes from this property name. I don't want to be confusing with the argument_type. Or should I still change it to type?

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1886,6 +1866,65 @@ def __init__(self, arg, arg_info, call):
+                return REAL_TYPE
+            if self.space.lower() == "go_i_scalar":
+                return INTEGER_TYPE
+            raise InternalError("GOcean expects scalar arguments to be in the"
```
Ok changed to type here.

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1926,7 +1965,7 @@ class GOKernelGridArgument(Argument):
     :raises GenerationError: if the grid property is not recognised.
```
Done

sergisiso commented:
@arporter  This is ready for the next review.

> I note that the Fortran generated for eg8 isn't quite complete - we don't declare integer :: i.

I forgot to update the PR description regarding this. I originally was doing everything here, but then decided to spit the PR in 2 (this one and #1087 ) as the 2 features are quite distinct.

However, I was missing the index arguments at the beginning of the kernel call that I added now in this PR.

> I think your choice of name psyir_expression is better than my _grid_property_reference although I realise they do slightly different things. Would it make sense for you to rename _grid_property_reference to _grid_property_psyir_expression (or similar) do you think?

Good suggestion. Done

> Please could you add any touched classes to an __all__ at the end of gocean1p0.

I added all gocean1p0 classes :)

arporter commented on src/psyclone/gocean1p0.py:
```diff
@@ -644,23 +643,27 @@ def add_bounds(bound_info):
-                      outer-stop:inner-start:inner-stop
+
+        > bound_info = offset-type:field-type:iteration-space:outer-start:
+        >             outer-stop:inner-start:inner-stop``
```
Spurious "``" at end of line and line breaks in the example code aren't rendering correctly.

sergisiso replied:
The ">>>" syntax is the closer I could get to render them as expected. Would you do it differently?

arporter replied:
No, that looks fine to me.

arporter reviewed (changes requested):
Looking good now - thanks very much for fixing all of those docstring issues.
There are just a few minor things and a naming issue for the routine that generates the PSyIR for the kernel Arguments.
All tests are OK, incl. with compilation.
Ditto the examples.
Reference guide builds without additional warnings.
All changes are pycodestyle/pylint clean.

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -2293,14 +2314,16 @@ def depth(self, index0, index1):
+                           index of the associated array. This value \
+                           must be between -1 and 1
+
+        :return int: the depth of the stencil in the specified direction.
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1743,6 +1746,18 @@ def raw_arg_list(self):
 
+    def psyir_arguments_list(self):
+        '''
+        :returns: the PSyIR expressions representing this Arguments list.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -2876,9 +2876,9 @@ def lower_to_language_level(self):
-        # Add arguments PSyIR expressions as children
-        for argument in self.arguments.args:
-            call_node.addchild(argument.psyir_expression())
+        for argument in self.arguments.psyir_arguments_list():
```
Updated to "psyir_expressions"

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3445,6 +3445,14 @@ def raw_arg_list(self):
+    @abc.abstractmethod
+    def psyir_arguments_list(self):
+        '''
+        :returns: the PSyIR expressions representing this Arguments list.
```
Done

sergisiso commented on src/psyclone/tests/gocean1p0_test.py:
```diff
@@ -1759,25 +1759,56 @@ def test_gokernelargument_infere_datatype():
+    assert (argument_list[1].symbol is
+            kernelcall.scope.symbol_table.lookup_with_tag(
+                "noncontiguous_kidx"))
 
```
Good idea

sergisiso commented:
@arporter Ready for next review

arporter reviewed (approved):
All requested changes have been made.
I've manually checked coverage and it is OK.
All tests OK with compilation. Ditto examples.
Documentation all builds cleanly.
